### PR TITLE
uoa: enable nf hook in each network namespace

### PR DESCRIPTION
**Pull request by lubinszARM**
QUIC-LoadBalancer@container needs this feature.

Signed-off-by: Robin Lu < lubinsz@gmail.com>
Signed-off-by: Robin Lu <robin.lubin@shopee.com>

### CODE REVIEW

**Comment by lubinszARM** [comment]
Hi @ywc689 
Any comments?

File: kmod/uoa/uoa.c [hunk 1]
```diff
@@ -922,6 +922,46 @@ static struct nf_hook_ops uoa_nf_hook_ops6[] __read_mostly = {
     },
 };
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,3,0)
```
Comment (ywc689) [inline kmod/uoa/uoa.c:925]: Can the feature support kernel version elder than 4.3.0?

**Comment by ywc689** [comment]
> QUIC-LoadBalancer@container needs this feature.
> 
> Signed-off-by: Robin Lu < [lubinsz@gmail.com](mailto:lubinsz@gmail.com)> Signed-off-by: Robin Lu [robin.lubin@shopee.com](mailto:robin.lubin@shopee.com)

Why the previous uoa can't support the case? Could you give a detail test case for the feature?

**Comment by lubinszARM** [comment]
> > QUIC-LoadBalancer@container needs this feature.
> > Signed-off-by: Robin Lu < [lubinsz@gmail.com](mailto:lubinsz@gmail.com)> Signed-off-by: Robin Lu [robin.lubin@shopee.com](mailto:robin.lubin@shopee.com)
> 
> Why the previous uoa can't support the case? Could you give a detail test case for the feature?

Plz see my test steps as reference:
1 测试方法
使用DPVS UOA模块自带的工具uperf client(模拟NLB发带IP option的包)和uperf server。
uperf client程序发包加上了uoa选项(IP option里的源IP为97.97.97.97，端口为55035)，
uperf server会打印收到的uoa信息(real client 97.97.97.97:55035)

2 topology
image

3 测试场景：

(1)1 SDN POD(NLB,uperf client模拟) - 1 SDN POD(ALB,uperf server模拟),NLB不断往ALB发带IP option的包


检查点:

(1)每个server端的pod应该有如下打印:
image
（2）抓包查看客户端发出的UDP包中有IP options属性.
image

**Comment by lubinszARM** [comment]
> > QUIC-LoadBalancer@container needs this feature.
> > Signed-off-by: Robin Lu < [lubinsz@gmail.com](mailto:lubinsz@gmail.com)> Signed-off-by: Robin Lu [robin.lubin@shopee.com](mailto:robin.lubin@shopee.com)
> 
> Why the previous uoa can't support the case? Could you give a detail test case for the feature?

Because , we deploy the uoa server in the container, not in the host.
The old driver only be hooked in host, not in each network namespace.

**Review by ywc689 — APPROVED** [review]

File: kmod/uoa/uoa.c [hunk 2]
```diff
@@ -939,6 +939,46 @@ static struct nf_hook_ops uoa_nf_hook_ops6[] __read_mostly = {
...
+static int __net_init __uoa_init(struct net *net)
+{
+    int ret = -ENOMEM;
+
+    ret = nf_register_net_hooks(net, uoa_nf_hook_ops, ARRAY_SIZE(uoa_nf_hook_ops));
+    if (ret < 0)
+        goto ops_hook_fail;
+
+    ret = nf_register_net_hooks(&init_net, uoa_nf_hook_ops6,
```
Comment (donghaobo) [inline kmod/uoa/uoa.c:951]: why init_net for ipv6?

Comment (lubinszARM) [reply kmod/uoa/uoa.c:951]: Ahhh. Sorry for late.
I just added the supporting for ipv6, but only tested the patch under ipv4.
So, if you don't like it. I can remove the supporting for ipv6.

Comment (donghaobo) [reply kmod/uoa/uoa.c:951]: > Ahhh. Sorry for late. I just added the supporting for ipv6, but only tested the patch under ipv4. So, if you don't like it. I can remove the supporting for ipv6.

i mean it should be `ret = nf_register_net_hooks(net, uoa_nf_hook_ops6,`

**Pull request merged**